README

				   GStreamer
				  -----------

This module allows you to use the GStreamer library from Perl.

GStreamer comes with automatically generated API documentation.  Its index can
be accessed with:

  perldoc GStreamer::index

				 Installation
				--------------

To install GStreamer, run the following commands:

  perl Makefile.PL
  make
  make test
  make install

				 Dependencies
				--------------

GStreamer depends on any release of the 0.10 series of the GStreamer library
and on the following Perl modules:

  ExtUtils::Depends
  ExtUtils::PkgConfig
  Glib >= 1.180


NEWS
Overview of changes in GStreamer 0.19
=====================================

* Avoid misusing the macro PL_na, thus preventing potential issues when
  GStreamer is used in conjunction with certain XS modules, among them
  XML::Parser and String::Approx.

Overview of changes in GStreamer 0.18
=====================================

* Add GStreamer::Buffer::data_ptr() to access the buffer's data without copying
  it. (RT#82826)
∗ Make a few tests more robust
* Created %meta_merge which follows v2 of CPAN Meta Spec
  - %meta_merge is used for generating metadata via
    ExtUtils::MakeMaker/CPAN::Meta

Overview of changes in GStreamer 0.17
=====================================

* Make GStreamer::Buffer not leak all data.
∗ Change FSF address in license notices (RT#72664/74533)

Overview of changes in GStreamer 0.16
=====================================

[The actual 0.16 release did not contain these notes due to an
 oversight.]

* Make the read-metadata.pl example more robust.
* Fix many test failures.

Overview of changes in GStreamer 0.15
=====================================

* Wrap gst_pad_is_blocking, gst_pad_peer_query, gst_pad_iterate_internal_links,
  and gst_pad_iterate_internal_links_default.
* Don't assume ownership of GstIndexEntry objects returned by the various
  GstIndex methods.
* Throw an exception if unknown GstFormat values are encountered.
* Convert NULL GstIterator objects to undef.
* Fix a few test failures.

Overview of changes in GStreamer 0.14
=====================================

* Fix a few test failures.
* Hush compiler warnigns.

Overview of changes in GStreamer 0.13
=====================================

* Make the return value of GStreamer::Element::add_pad,
  GStreamer::Element::remove_pad, and GStreamer::Element::set_locked_state
  meaningful.
* Fix a few test and build issues.

Overview of changes in GStreamer 0.12
=====================================

* Add support for GST_TYPE_FOURCC, GST_TYPE_DOUBLE_RANGE, GST_TYPE_ARRAY,
  GST_TYPE_FRACTION, and GST_TYPE_FRACTION_RANGE.

Overview of changes in GStreamer 0.11
=====================================

* Fix build and test suite issues.

Overview of changes in GStreamer 0.10
=====================================

* Alter GStreamer::Structure::from_string to return only one value.
* Add support for latency, async-start, and async-done messages.
* Add GStreamer::Message::timestamp() and src().
* Fix test failures.

Overview of changes in GStreamer 0.09
=====================================

* Fix a portability issue.  [Jeremy Messenger]
* Fix a few test suite issues.

Overview of changes in GStreamer 0.08
=====================================

* Fix a few more test suite issues.

Overview of changes in GStreamer 0.07
=====================================

* Make GStreamer::Clock::wait_async work correctly.
* Fix some test suite issues.

Overview of changes in GStreamer 0.06
=====================================

* Fix an obscure GStreamer::Query bug.
* Add handlers for GstDate.  They convert to and from seconds since epoch.
* Require Glib 1.114 for the type-to-package mapping changes.  This greatly
  improves support for interfaces like GStreamer::TagSetter.

Overview of changes in GStreamer 0.05
=====================================

* Port to GStreamer 0.10.0.
  - Wrap and test most of the new API.
  - Remove old stuff.
* Use Glib's 64-bit integer converters.
* Rename the various GStreamer::Caps and GStreamer::Event constructors from
  new_thing to GStreamer::Caps::Thing::new and GStreamer::Event::Thing::new
  respectively.

Overview of changes in GStreamer 0.04
=====================================

* Get rid of the Gtk2 dependency by using Glib 1.100's Glib::CodeGen.

Overview of changes in GStreamer 0.03
=====================================

* Implement overloaded operators for GStreamer::Caps: +, -, &, |, <=, >=, ==
  and "".
* Fix a few test failures.

Overview of changes in GStreamer 0.02
=====================================

* Fix handling of 64bit integers all over the place.
* Fix the processing of GStreamer::Element::seek's seek-type argument.
* Fix the "location" property to not mangle filenames.
* Add support for tag reading and writing.
* Add plugin and library loading support.

Overview of changes in GStreamer 0.01
=====================================

* Initial release.


Makefile.PL
#!/usr/bin/perl
use 5.008;
use strict;
use warnings;

use ExtUtils::MakeMaker;
use Cwd qw(cwd);

# $Id$

our %build_reqs = (
  'perl-ExtUtils-Depends'   => '0.205',
  'perl-ExtUtils-PkgConfig' => '1.07',
  'perl-Glib'               => '1.180',
  'libgstreamer'            => '0.10.0',
);

our %pre_reqs = (
  'Glib'                => $build_reqs{'perl-Glib'},
  'ExtUtils::Depends'   => $build_reqs{'perl-ExtUtils-Depends'},
  'ExtUtils::PkgConfig' => $build_reqs{'perl-ExtUtils-PkgConfig'},
);

unless (eval "use ExtUtils::Depends '$build_reqs{'perl-ExtUtils-Depends'}';"
           . "use ExtUtils::PkgConfig '$build_reqs{'perl-ExtUtils-PkgConfig'}';"
           . "use Glib '$build_reqs{'perl-Glib'}';"
           . "use Glib::MakeHelper;"
           . "use Glib::CodeGen;"
           . "1") {
  warn "$@\n";
  WriteMakefile(
    PREREQ_FATAL => 1,
    PREREQ_PM    => \%pre_reqs,
  );
  exit 1; # not reached
}

# If the package can't be found, warn and exit with status 0 to indicate to
# CPAN testers that their system is not supported.
my %pkgcfg;
unless (eval { %pkgcfg = ExtUtils::PkgConfig->find (
                 "gstreamer-0.10 >= $build_reqs{'libgstreamer'}");
	       1; })
{
	warn $@;
	exit 0;
}

mkdir 'build', 0777;

our @xs_files = <xs/*.xs>;
our %pod_files = (
	'lib/GStreamer.pm' => '$(INST_MAN3DIR)/GStreamer.$(MAN3EXT)',
	Glib::MakeHelper->do_pod_files (@xs_files),
);

# --------------------------------------------------------------------------- #

sub gen_object_stuff {
    my ($typemacro, $classname, $root, $package) = @_;

    Glib::CodeGen::add_typemap("$classname *", "T_GPERL_GENERIC_WRAPPER");
    Glib::CodeGen::add_typemap("const $classname *", "T_GPERL_GENERIC_WRAPPER");
    Glib::CodeGen::add_typemap("$classname\_ornull *", "T_GPERL_GENERIC_WRAPPER");
    Glib::CodeGen::add_typemap("const $classname\_ornull *", "T_GPERL_GENERIC_WRAPPER");

    Glib::CodeGen::add_register(<<"EOT");
#ifdef $typemacro
gperl_register_object ($typemacro, "$package");
#endif /* $typemacro */
EOT

    Glib::CodeGen::add_header(<<"EOT");
#ifdef $typemacro
  typedef $classname $classname\_ornull;
# define Sv$classname(sv)		(($classname *) gperl_get_object_check (sv, $typemacro))
# define newSV$classname(val)		(gperl_new_object (G_OBJECT (val), TRUE))
# define Sv$classname\_ornull(sv)	(((sv) && SvOK (sv)) ? Sv$classname (sv) : NULL)
# define newSV$classname\_ornull(val)	(((val) == NULL) ? &PL_sv_undef : newSV$classname (val))
#endif /* $typemacro */
EOT
}

sub gen_mini_object_stuff {
    my ($typemacro, $classname, $root, $package) = @_;

    Glib::CodeGen::add_typemap("$classname *", "T_GPERL_GENERIC_WRAPPER");
    Glib::CodeGen::add_typemap("const $classname *", "T_GPERL_GENERIC_WRAPPER");
    Glib::CodeGen::add_typemap("$classname\_ornull *", "T_GPERL_GENERIC_WRAPPER");
    Glib::CodeGen::add_typemap("const $classname\_ornull *", "T_GPERL_GENERIC_WRAPPER");
    Glib::CodeGen::add_typemap("$classname\_noinc *", "T_GPERL_GENERIC_WRAPPER");
    Glib::CodeGen::add_typemap("const $classname\_noinc *", "T_GPERL_GENERIC_WRAPPER");
    Glib::CodeGen::add_typemap("$classname\_noinc_ornull *", "T_GPERL_GENERIC_WRAPPER");
    Glib::CodeGen::add_typemap("const $classname\_noinc_ornull *", "T_GPERL_GENERIC_WRAPPER");

    Glib::CodeGen::add_register(<<"EOT");
#ifdef $typemacro
gst2perl_register_mini_object ($typemacro, "$package");
#endif /* $typemacro */
EOT

    Glib::CodeGen::add_header(<<"EOT");
#ifdef $typemacro
  typedef $classname $classname\_ornull;
  typedef $classname $classname\_noinc;
  typedef $classname $classname\_noinc_ornull;
# define Sv$classname(sv)			(($classname *) gst2perl_mini_object_from_sv (sv))
# define Sv$classname\_ornull(sv)		(((sv) && SvOK (sv)) ? Sv$classname(sv) : NULL)
# define newSV$classname(val)			(gst2perl_sv_from_mini_object (GST_MINI_OBJECT (val), TRUE))
# define newSV$classname\_ornull(val)		(((val) == NULL) ? &PL_sv_undef : newSV$classname(val))
# define newSV$classname\_noinc(val)		(gst2perl_sv_from_mini_object (GST_MINI_OBJECT (val), FALSE))
# define newSV$classname\_noinc_ornull(val)	(((val) == NULL) ? &PL_sv_undef : newSV$classname\_noinc(val))
#endif /* $typemacro */
EOT
}

Glib::CodeGen->add_type_handler (GstObject => \&gen_object_stuff);
Glib::CodeGen->add_type_handler (GstMiniObject => \&gen_mini_object_stuff);

Glib::CodeGen->parse_maps('gst2perl');
Glib::CodeGen->write_boot(ignore => qr/^GStreamer$/);

# --------------------------------------------------------------------------- #

ExtUtils::PkgConfig->write_version_macros (
  "build/gst2perl-version.h",
  "gstreamer-0.10" => "GST",
);

my $gst = ExtUtils::Depends->new('GStreamer', 'Glib');
$gst->set_inc($pkgcfg{cflags} . ' -I./build ');
$gst->set_libs($pkgcfg{libs});
$gst->add_xs(@xs_files);
$gst->add_pm('lib/GStreamer.pm' => '$(INST_LIBDIR)/GStreamer.pm');
$gst->add_typemaps(map {File::Spec->catfile(cwd(), $_)}
                       ('gst.typemap', 'build/gst2perl.typemap'));

$gst->install(qw(gst2perl.h
                 build/gst2perl-autogen.h
                 build/gst2perl-version.h
                 doctypes));
$gst->save_config('build/IFiles.pm');

WriteMakefile(
  NAME          => 'GStreamer',
  VERSION_FROM  => 'lib/GStreamer.pm',
  ABSTRACT_FROM => 'lib/GStreamer.pm',
  PREREQ_PM     => \%pre_reqs,
  XSPROTOARG    => '-noprototypes',
  MAN3PODS      => \%pod_files,

  META_MERGE    => {
    configure_requires => \%pre_reqs,
    no_index           => { directory => [qw/xs/] },
  },

  $gst->get_makefile_vars,
);

sub MY::postamble {
  return Glib::MakeHelper->postamble_clean ()
       . Glib::MakeHelper->postamble_docs_full (
           DEPENDS => $gst,
           DOCTYPES => 'doctypes',
           COPYRIGHT_FROM => 'copyright.pod');
}
